src/UCTNode.cpp

```diff
@@ -412,133 +349,75 @@ UCTNode* UCTNode::uct_select_child(int color) {
 
         if (value > best_value) {
             best_value = value;
-            best = child;
-        }
-
-        child = child->m_nextsibling;
-        // Make sure we are at a valid successor.
-        while (child != nullptr && !child->valid()) {
-            child = child->m_nextsibling;
+            best = child.get();
         }
     }
 
     assert(best != nullptr);
-
     return best;
 }
 
-class NodeComp : public std::binary_function<UCTNode::sortnode_t,
-                                             UCTNode::sortnode_t, bool> {
+class NodeComp : public std::binary_function<std::unique_ptr<UCTNode>&,
+                                             std::unique_ptr<UCTNode>&, bool> {
 public:
-    NodeComp() = default;
-    // winrate, visits, score, child
-    //        0,     1,     2,     3
-
-    bool operator()(const UCTNode::sortnode_t a, const UCTNode::sortnode_t b) {
-        // One node has visits, the other does not
-        if (!std::get<1>(a) && std::get<1>(b)) {
-            return false;
-        }
-
-        if (!std::get<1>(b) && std::get<1>(a)) {
-            return true;
+    NodeComp(int color) : m_color(color) {};
```

Comment:
Should an `explicit` constructor be used to prevent implicit conversion?

Reply:
My understanding from http://en.cppreference.com/w/cpp/language/explicit
is that adding explicit will prevent copy initialization (which would be fine in this case).

but I defer to a c++ expert (you or gcp)